### ISSUE TITLE
Move TorchInductor Triton autotuner to compile time, use common cache

### ISSUE THREAD

**Issue by jansel** [pytorch/torchdynamo#824]
We use the Triton autotuner here:
https://github.com/pytorch/torchdynamo/blob/8e4abe3f66fb6b8d97d6461d64f9c3be323a523e/torchinductor/triton_ops/autotune.py#L182
and, here:
https://github.com/pytorch/torchdynamo/blob/8e4abe3f66fb6b8d97d6461d64f9c3be323a523e/torchinductor/triton_ops/autotune.py#L216
To select tile sizes for generated kernels from a short list of options.  

This results in good performance, but has a high cost in terms of warm up time.  We will be running nearly identical autotuning runs to pick tilings that should be obvious.

I'd propose we create a small cost model to replace the autotuner.  As an input to this model we should take:
1) The size_hints (already rounded up to next power of two)
2) The indexing formulas of all the reads and writes
3) [Maybe] Number of `tmpX` variables used in kernel
4) [Maybe] Number of reduction loops in the kernel
5) [Maybe] Histogram of `ops.*` usage in the kernel

As an initial implementation, we could just build an on-disk cache/hash-table that runs the Triton autotuner on misses, but reuses a prior decision if the model inputs are identical.  This alone should dramatically speed up warmup times.

Longer term, we could take the data collected in this cache and train a small neural network to predict which tiling to use.  For this we should have the neural network predict the normalized runtime for each tiling choice.  This cost model can then replace the measurement step of the autotuner.

**Comment by bankawas** [comment]
Hi @jansel, I think this is a great idea:

> Create a small cost model to replace the autotuner
> As an initial implementation, we could just build an on-disk cache/hash-table that runs the Triton autotuner on misses, but reuses a prior decision if the model inputs are identical.

I have a few questions/suggestions:
- for on-disk cache/hash-table;
  - is it dynamically updated? how big can it get? 
  - if we dynamically update it, how do we go about that? What qualifies as a good entry to the hash-table from "prior decisions"? Is it based on frequency analysis of repeated input/decision pairs? 

- Would it be meaningful to train a binary classifier that predicts misses?

- How about training a classifier to predict the right tiling choice? 


Thank you for answering my question via chat on runtime, pasting it here for reference:

> Longer term, we could take the data collected in this cache and train a small neural network to predict which tiling to use. For this we should have the neural network predict the normalized runtime for each tiling choice. This cost model can then replace the measurement step of the autotuner.
- is the normalized runtime for each tiling deterministic?

@jansel answer: It should be deterministic given the same sizes + kernel + hardware.  Though measurement is noisy, so what the autotuner does is run it many times and averages them. We are currently having issues with compile time because of that.  Doing this is expensive.

A follow up question on runtime determinism + on-disk hash-table idea:

> deterministic given the same sizes + kernel + hardware.

Does that mean we want to dynamically populate the hash-table on-disk, and ultimately, train a neural network to predict tiling choice for the same sizes+kernel+hardware?

**Comment by bankawas** [comment]
@sangongs shared this awesome resource: https://github.com/merrymercy/awesome-tensor-compilers#cost-model 

@jansel which paper would you recommend looking into first?

**Comment by jansel** [comment]
>     * for on-disk cache/hash-table;
>       
>       * is it dynamically updated? how big can it get?
>       * if we dynamically update it, how do we go about that? What qualifies as a good entry to the hash-table from "prior decisions"? Is it based on frequency analysis of repeated input/decision pairs?

Yes, it should be dynamically updated.

Cache entries should be small.  Much smaller than the current TorchInductor code cache, which is unbounded. So I wouldn't worry too much about size.  If it becomes an issue we can add a cap.
 
>     * Would it be meaningful to train a binary classifier that predicts misses?

Couldn't you check for a miss in `O(1)` time with a hash table?  How would this be used?
 
>     * How about training a classifier to predict the right tiling choice?

Yes this seems useful.

I would train a supervised learning classifier that takes properties of the kernel (mostly memory addresses and sizes) as features, and predicts the normalized execution time of each kernel.  You could get this training data from the code cache.  You could then replace running the kernels with that model.

> Does that mean we want to dynamically populate the hash-table on-disk, and ultimately, train a neural network to predict tiling choice for the same sizes+kernel+hardware?

Yes, that is correct.

> @jansel which paper would you recommend looking into first?

The [Triton Paper](http://www.eecs.harvard.edu/~htk/publication/2019-mapl-tillet-kung-cox.pdf) would be a good start.

**Comment by bankawas** [comment]
Thank you, @jansel 🙂

> Couldn't you check for a miss in O(1) time with a hash table? How would this be used?

Yeah, of course. I think a simple lookup is best if the table doesn't get too big. 

I wasn't sure how we update the hash-table (my question below), or how big it would get. 

A binary classifier may be meaningful to avoid space complexity. It may also be useful if the dynamic update of the hash-table is more complex than simply adding entries based on occurrence. 

> * if we dynamically update it, how do we go about that? What qualifies as a good entry to the hash-table from "prior decisions"? Is it based on frequency analysis of repeated input/decision pairs?
